linkerd/main/src/main/scala/io/buoyant/linkerd/Main.scala

```diff
@@ -34,12 +34,14 @@ object Main extends App {
         val admin = initAdmin(config, linker)
         val telemeters = linker.telemeters.map(_.run())
         val routers = linker.routers.map(initRouter(_))
+        val closableNamers = linker.namers.map(_._2).collect { case x: Closable => x }
 
         log.info("linkerd initialized.")
         registerTerminationSignalHandler(config.admin.flatMap(_.shutdownGraceMs))
         closeOnExit(Closable.sequence(
           Closable.all(routers: _*),
           Closable.all(telemeters: _*),
+          Closable.all(closableNamers: _*),
```

Comment:
Since the close returns a future there is a wait further down in this function which waits on those futures.
        Await.all(routers: _*)
        Await.all(telemeters: _*)

which will need to be changed to 
        Await.all(closableNamers: _*)
        Thread.sleep(<zookeeper session timeout of clients>)
        Await.all(routers: _*)
        Await.all(telemeters: _*)

since the first thing we probably want to do is un-register ourselves from zookeeper. We will then need to wait for the default zookeeper session timeout for this change to propagate to the clients.